src/proto/streams/counts.rs

```diff
@@ -122,10 +151,20 @@ impl Counts {
     {
         // TODO: Does this need to be computed before performing the action?
         let is_pending_reset = stream.is_pending_reset_expiration();
+        let was_reserved = stream.state.is_reserved_remote();
 
         // Run the action
         let ret = f(self, &mut stream);
 
+        // If the reserved state has changed, adjust counter accordingly
+        if stream.state.is_reserved_remote() != was_reserved {
+            if was_reserved {
+                self.dec_num_reserved_streams();
+            } else {
+                self.inc_num_reserved_streams();
+            }
+        }
+
```

Comment:
I don't feel entirely comfortable about this: it's on quite a hot path (every single state transition) for relatively rare event (PUSH_PROMISE frame received and subsequent HEADER frame received)—but the alternatives seem a bit hacky too.  I'm going to take a closer look at how state transitions are verified and performed, as it feels like it may benefit from some refactoring.